File: gems/decomplex/lib/decomplex/root_cause.rb
# frozen_string_literal: true

require_relative "convergence"

module Decomplex
  # Root-cause clustering. Convergence answers "which (file, method) is
  # implicated by the most detectors" (a LOCATION hotspot). RootCause
  # answers the question that actually drives a rearchitecture: "which
  # ONE underlying entity, when fixed, collapses a cluster of findings
  # across detectors" -- the 74-findings-are-really-one-invariant view
  # (e.g. the `storage`/`provenance` co-write surface that
  # Neglected-Updates + Derived-State + Decision-Pressure + False-
  # Simplicity all name independently).
  #
  # Pure READ-ONLY consumer of Report#sections_data, exactly like
  # Convergence: it re-derives no detector. It only re-groups their
  # already-published findings by the ENTITY they are about instead of
  # by their location, and names the single fix shape.
  #
  # Two entity kinds (a finding contributes whichever it carries):
  #   :name  -- a shared identifier (attribute/contract/predicate). The
  #             cross-detector "same state" link (storage/provenance).
  #   :tuple -- a shared decision tuple (a guard/dispatch member set).
  #             Links the dispatch/path detectors on the same decision.
  class RootCause
    # Generic field families (NOT per-detector logic -- the same three
    # buckets every detector's findings already use).
    TUPLE_FIELDS = %i[members guards pattern].freeze
    NAME_ARRAY_FIELDS = %i[pair names].freeze
    NAME_STR_FIELDS = %i[field derived source contract canon predicate
                         detail ref_name has missing].freeze
    STOPWORDS = %w[nil true false self end do if then else self_ it
                   new to_s call each map].freeze

    # Ordered most-specific-first. Maps the SET of detectors that
    # converged to the one fix shape (SlopCop-style interpretation
    # policy: decomplex owns findings; this only names the action,
    # ranked-candidate, never a verdict).
    FIX_SHAPE = [
      [%w[Neglected\ Updates Derived-State\ Staleness], :name,
       "single-source this state (one stamp, or recompute on write) -- the invariant-#16 desync shape"],
      [%w[Broken\ Protocols], :any,
       "pair the protocol (RAII / ensure); the unpaired site is the deviant"],
      [%w[Missing\ Abstractions Reification\ Misses Semantic\ Predicate\ Aliases Exact\ Predicate\ Aliases], :any,
       "reify ONE named predicate/decision and call it everywhere"],
      [%w[Missing\ Abstractions Neglected\ Conditions Neglected\ Path\ Conditions], :tuple,
       "extract the decision; if it dispatches a closed set, consider product-vs-sum (fat-union -> nil-kill)"],
      [%w[Decision\ Pressure], :any,
       "tighten the contract once; the scattered defensive guards collapse (cross-proc -> nil-kill)"]
    ].freeze

    # sections: [[title, tier, findings], ...] (Report#sections_data).
    def self.cluster(sections, min_detectors: 2)
      tier_of = {}
      acc = Hash.new do |h, k|
        h[k] = { dets: {}, findings: [], tiers: {} }
      end
      sections.each do |title, tier, findings|
        next unless findings

        tier_of[title] = tier
        findings.each do |f|
          entities(f).each do |ent|
            u = acc[ent]
            u[:dets][title] = true
            u[:tiers][title] = tier
            u[:findings] << f
          end
        end
      end

      acc.filter_map do |(kind, token), u|
        next if u[:dets].size < min_detectors

        dets = u[:dets].keys.sort
        units = u[:findings].flat_map { |f| finding_units(f) }.uniq
        score = u[:tiers].values.sum { |t| Convergence::TIER_WEIGHT.fetch(t, 1) }
        fat = fat_union?(kind, token, u[:findings])
        { kind: kind, token: token, detectors: dets,
          n_detectors: dets.size, support: u[:findings].size,
          scatter: units.size, score: score, fat_union: fat,
          fix: fat ? FAT_UNION_FIX : fix_shape(dets, kind),
          sites: u[:findings].flat_map { |f| Convergence.locations(f) }
                             .uniq.first(8) }
      end.sort_by do |h|
        [-h[:n_detectors], -h[:score], -h[:scatter], h[:kind].to_s, h[:token]]
      end
    end

    FAT_UNION_FIX =
      "fat union -- decompose product-vs-sum: hoist the common fields " \
      "to a struct, keep a SMALL union for the variant part " \
      "(extraction is value-object work -> nil-kill owns it)"

    CONST_SEG = /\A(::)?[A-Z]\w*(::[A-Z]\w*)*\z/.freeze

    # #2 precursor: a :tuple cluster whose dispatch is over CLASS
    # CONSTANTS (not symbols/ints) and is a `case` dispatch (Missing
    # Abstractions carries kind: :case_dispatch) is a union being
    # re-dispatched -- the fat-union signature, surfaced without a
    # dedicated detector. Constant check excludes enum-ish int/symbol
    # dispatch (that is not a union).
    def self.fat_union?(kind, token, findings)
      return false unless kind == :tuple
      return false unless findings.any? { |f| f[:kind] == :case_dispatch }

      members = token.split(" | ")
      members.size >= 2 && members.all? { |m| m =~ CONST_SEG }
    end

    # -> Set of [kind, token]. Generic over the field families; no
    # per-detector branching.
    def self.entities(finding)
      out = []
      TUPLE_FIELDS.each do |k|
        v = finding[k]
        next unless v.is_a?(Array) && v.size >= 2

        # NO field-name prefix: Missing-Abstractions `:members`,
        # Neglected-Conditions `:pattern` and PathCondition `:guards`
        # for the SAME decision must yield the SAME token (that cross-
        # detector link is the whole point of a tuple-entity).
        out << [:tuple, v.map(&:to_s).sort.join(" | ")[0, 160]]
      end
      NAME_ARRAY_FIELDS.each do |k|
        v = finding[k]
        next unless v.is_a?(Array)

        v.each { |e| tokens(e).each { |t| out << [:name, t] } }
      end
      NAME_STR_FIELDS.each do |k|
        v = finding[k]
        tokens(v).each { |t| out << [:name, t] } if v
      end
      out.uniq
    end

    # Bare identifier tokens: strip a leading @/:/. and trailing =?!,
    # drop trivial/keyword/short noise. "storage=" / ".storage" /
    # "@storage" / "storage" all -> "storage" (the cross-detector link).
    def self.tokens(val)
      val.to_s.scan(/[A-Za-z_][A-Za-z0-9_]*[?!=]?/).filter_map do |w|
        t = w.sub(/[?!=]\z/, "")
        next if t.length < 2 || STOPWORDS.include?(t)

        t
      end.uniq
    end

    def self.finding_units(finding)
      Convergence.locations(finding).filter_map do |loc|
        file, meth, = Convergence.parse_loc(loc)
        [file, meth] if file && meth
      end
    end

    def self.fix_shape(detectors, kind)
      FIX_SHAPE.each do |titles, want_kind, label|
        next unless want_kind == :any || want_kind == kind
        next if (titles & detectors).empty?

        return label
      end
      "converging structural debt -- resolve once at the named entity"
    end
  end
end

File: gems/decomplex/rust/src/decomplex/root_cause.rs
use crate::decomplex::convergence;
use crate::decomplex::report::ReportSection;
use crate::decomplex::report_value as rv;
use regex::Regex;
use serde_json::Value;
use std::collections::{BTreeMap, HashMap, HashSet};
use std::sync::OnceLock;

const TUPLE_FIELDS: &[&str] = &["members", "guards", "pattern"];
const NAME_ARRAY_FIELDS: &[&str] = &["pair", "names"];
const NAME_STR_FIELDS: &[&str] = &[
    "field",
    "derived",
    "source",
    "contract",
    "canon",
    "predicate",
    "detail",
    "ref_name",
    "has",
    "missing",
];
const STOPWORDS: &[&str] = &[
    "nil", "true", "false", "self", "end", "do", "if", "then", "else", "self_", "it", "new",
    "to_s", "call", "each", "map",
];
const FAT_UNION_FIX: &str = "fat union -- decompose product-vs-sum: hoist the common fields to a struct, keep a SMALL union for the variant part (extraction is value-object work -> nil-kill owns it)";

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct Entity {
    pub kind: String,
    pub token: String,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct Cluster {
    pub kind: String,
    pub token: String,
    pub detectors: Vec<String>,
    pub n_detectors: usize,
    pub support: usize,
    pub scatter: usize,
    pub score: i64,
    pub fat_union: bool,
    pub fix: String,
    pub sites: Vec<String>,
}

#[derive(Clone, Debug)]
struct Accumulator {
    dets: BTreeMap<String, bool>,
    findings: Vec<Value>,
    tiers: BTreeMap<String, i64>,
}

pub fn cluster(sections: &[ReportSection], min_detectors: usize) -> Vec<Cluster> {
    let mut acc: HashMap<Entity, Accumulator> = HashMap::new();
    for section in sections {
        for finding in &section.findings {
            for entity in entities(finding) {
                let row = acc.entry(entity).or_insert_with(|| Accumulator {
                    dets: BTreeMap::new(),
                    findings: Vec::new(),
                    tiers: BTreeMap::new(),
                });
                row.dets.insert(section.title.clone(), true);
                row.tiers.insert(section.title.clone(), section.tier);
                row.findings.push(finding.clone());
            }
        }
    }

    let mut clusters = acc
        .into_iter()
        .filter_map(|(entity, row)| {
            if row.dets.len() < min_detectors {
                return None;
            }
            let detectors = row.dets.keys().cloned().collect::<Vec<_>>();
            let mut units = row
                .findings
                .iter()
                .flat_map(finding_units)
                .collect::<Vec<_>>();
            units.sort();
            units.dedup();
            let score = row
                .tiers
                .values()
                .map(|tier| convergence::tier_weight(*tier))
                .sum();
            let fat_union = fat_union(&entity.kind, &entity.token, &row.findings);
            let mut sites = row
                .findings
                .iter()
                .flat_map(convergence::locations)
                .collect::<Vec<_>>();
            let mut seen_sites = HashSet::new();
            sites.retain(|site| seen_sites.insert(site.clone()));
            sites.truncate(8);
            Some(Cluster {
                kind: entity.kind.clone(),
                token: entity.token.clone(),
                n_detectors: detectors.len(),
                support: row.findings.len(),
                scatter: units.len(),
                score,
                fat_union,
                fix: if fat_union {
                    FAT_UNION_FIX.to_string()
                } else {
                    fix_shape(&detectors, &entity.kind)
                },
                detectors,
                sites,
            })
        })
        .collect::<Vec<_>>();
    clusters.sort_by(|left, right| {
        right
            .n_detectors
            .cmp(&left.n_detectors)
            .then_with(|| right.score.cmp(&left.score))
            .then_with(|| right.scatter.cmp(&left.scatter))
            .then_with(|| left.kind.cmp(&right.kind))
            .then_with(|| left.token.cmp(&right.token))
    });
    clusters
}

pub fn entities(finding: &Value) -> Vec<Entity> {
    let mut out = Vec::new();
    for key in TUPLE_FIELDS {
        let values = rv::array(finding, key);
        if values.len() < 2 {
            continue;
        }
        let mut members = values
            .iter()
            .map(|value| rv::string(Some(value)))
            .collect::<Vec<_>>();
        members.sort();
        out.push(Entity {
            kind: "tuple".to_string(),
            token: truncate_chars(&members.join(" | "), 160),
        });
    }
    for key in NAME_ARRAY_FIELDS {
        for value in rv::array(finding, key) {
            for token in tokens(&rv::string(Some(value))) {
                out.push(Entity {
                    kind: "name".to_string(),
                    token,
                });
            }
        }
    }
    for key in NAME_STR_FIELDS {
        if let Some(value) = rv::get(finding, key) {
            for token in tokens(&rv::string(Some(value))) {
                out.push(Entity {
                    kind: "name".to_string(),
                    token,
                });
            }
        }
    }
    let mut seen = HashSet::new();
    out.retain(|entity| seen.insert((entity.kind.clone(), entity.token.clone())));
    out
}

pub fn tokens(value: &str) -> Vec<String> {
    static TOKEN_RE: OnceLock<Regex> = OnceLock::new();
    let re = TOKEN_RE.get_or_init(|| Regex::new(r"[A-Za-z_][A-Za-z0-9_]*[?!=]?").unwrap());
    let mut out = re
        .find_iter(value)
        .filter_map(|mat| {
            let token = mat.as_str().trim_end_matches(['?', '!', '=']).to_string();
            if token.len() < 2 || STOPWORDS.contains(&token.as_str()) {
                None
            } else {
                Some(token)
            }
        })
        .collect::<Vec<_>>();
    out.sort();
    out.dedup();
    out
}

pub fn finding_units(finding: &Value) -> Vec<(String, String)> {
    convergence::locations(finding)
        .into_iter()
        .filter_map(|loc| {
            let (file, method, _) = convergence::parse_loc(&loc);
            match (file, method) {
                (Some(file), Some(method)) => Some((file, method)),
                _ => None,
            }
        })
        .collect()
}

fn fat_union(kind: &str, token: &str, findings: &[Value]) -> bool {
    static CONST_RE: OnceLock<Regex> = OnceLock::new();
    let re = CONST_RE.get_or_init(|| Regex::new(r"\A(::)?[A-Z]\w*(::[A-Z]\w*)*\z").unwrap());
    if kind != "tuple" {
        return false;
    }
    if !findings
        .iter()
        .any(|finding| rv::kind_is(finding, "kind", "case_dispatch"))
    {
        return false;
    }
    let members = token.split(" | ").collect::<Vec<_>>();
    members.len() >= 2 && members.iter().all(|member| re.is_match(member))
}

fn fix_shape(detectors: &[String], kind: &str) -> String {
    let detectors = detectors.iter().map(String::as_str).collect::<HashSet<_>>();
    let shapes: &[(&[&str], &str, &str)] = &[
        (
            &["Neglected Updates", "Derived-State Staleness"],
            "name",
            "single-source this state (one stamp, or recompute on write) -- the invariant-#16 desync shape",
        ),
        (
            &["Broken Protocols"],
            "any",
            "pair the protocol (RAII / ensure); the unpaired site is the deviant",
        ),
        (
            &[
                "Missing Abstractions",
                "Reification Misses",
                "Semantic Predicate Aliases",
                "Exact Predicate Aliases",
            ],
            "any",
            "reify ONE named predicate/decision and call it everywhere",
        ),
        (
            &["Missing Abstractions", "Neglected Conditions", "Neglected Path Conditions"],
            "tuple",
            "extract the decision; if it dispatches a closed set, consider product-vs-sum (fat-union -> nil-kill)",
        ),
        (
            &["Decision Pressure"],
            "any",
            "tighten the contract once; the scattered defensive guards collapse (cross-proc -> nil-kill)",
        ),
    ];
    for (titles, want_kind, label) in shapes {
        if *want_kind != "any" && *want_kind != kind {
            continue;
        }
        if titles.iter().any(|title| detectors.contains(title)) {
            return (*label).to_string();
        }
    }
    "converging structural debt -- resolve once at the named entity".to_string()
}

fn truncate_chars(value: &str, max: usize) -> String {
    value.chars().take(max).collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn equivalent_state_tokens_collapse_to_same_name() {
        assert_eq!(tokens("@storage="), vec!["storage"]);
        assert_eq!(tokens(".storage"), vec!["storage"]);
    }

    #[test]
    fn tuple_fields_share_the_same_token() {
        let left = entities(&json!({"members": ["b", "a"]}));
        let right = entities(&json!({"guards": ["a", "b"]}));
        assert_eq!(left, right);
    }
}
